core/shared/src/main/scala/cats/effect/IO.scala

```diff
@@ -101,13 +121,50 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
 
   def race[B](that: IO[B]): IO[Either[A, B]] =
     racePair(that).flatMap {
-      case Left((a, f)) => f.cancel.as(Left(a))
-      case Right((f, b)) => f.cancel.as(Right(b))
+      case Left((oc, f)) =>
+        oc match {
+          case Outcome.Completed(fa) => f.cancel *> fa.map(Left.apply)
```

Comment:
```suggestion
          case Outcome.Completed(fa) => f.cancel *> fa.map(Left(_))
```

And analogously for the others. I think it's just a bit cleaner.